background_isolate_channels/lib/main.dart

```diff
@@ -0,0 +1,140 @@
+
+import 'simple_database.dart';
+
+///////////////////////////////////////////////////////////////////////////////
+// This is the UI which will present the contents of the [SimpleDatabase]. To
+// see where Background Isolate Channels are used see simple_database.dart.
+///////////////////////////////////////////////////////////////////////////////
+
+void main() {
+  runApp(const MyApp());
+}
+
+class MyApp extends StatelessWidget {
+  const MyApp({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: 'Background Isolate Channels',
+      theme: ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      home: const MyHomePage(title: 'Background Isolate Channels'),
+    );
+  }
+}
+
+class MyHomePage extends StatefulWidget {
+  const MyHomePage({super.key, required this.title});
+
+  final String title;
+
+  @override
+  State<MyHomePage> createState() {
+    return _MyHomePageState();
+  }
+}
+
+class _MyHomePageState extends State<MyHomePage> {
+  /// The database that is running on a background [Isolate]. This is nullable
+  /// because acquiring a [SimpleDatabase] is an asynchronous operation. This
+  /// value is `null` until the database is initialized.
+  SimpleDatabase? _database;
+
+  /// Local cache of the query results returned by the [SimpleDatabase] for the
+  /// UI to render from. It is nullable since querying the results is
+  /// asynchronous. The value is `null` before any result has been received.
+  List<String>? _entries;
+
+  /// What is searched for in the [SimpleDatabase].
+  String _query = '';
+
+  _MyHomePageState() {
+    // Write the value to [SharedPreferences] which will get read on the
+    // [SimpleDatabase]'s isolate.
+    final Future<void> sharedPreferencesSet = SharedPreferences.getInstance()
+        .then(
+            (sharedPreferences) => sharedPreferences.setBool('isDebug', true));
+    final Future<Directory> tempDirFuture =
+        path_provider.getTemporaryDirectory();
+
+    // Wait until the [SharedPreferences] value is set and the temporary
+    // directory is received before opening the database. If
+    // [sharedPreferencesSet] does not happen before opening the
+    // [SimpleDatabase] there has to be a way to refresh
+    // [_SimpleDatabaseServer]'s [SharedPreferences] cached values.
+    Future.wait([sharedPreferencesSet, tempDirFuture])
+        .then((List<dynamic> values) {
+      final Directory? tempDir = values[1];
+      final String dbPath = path.join(tempDir!.path, 'database.db');
+      SimpleDatabase.open(dbPath).then((SimpleDatabase database) {
+        setState(() {
+          _database = database;
+          _refresh();
+        });
+      });
+    });
+  }
+
+  /// Performs a find on [SimpleDatabase] with [query] and updates the listed
+  /// contents.
+  void _refresh({String? query}) {
+    if (query != null) {
+      _query = query;
+    }
+    _database?.find(_query).toList().then((entries) {
```

Comment:
Streams are somewhat difficult to deal with and we keep them out of the framework, see also https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-of-streams-in-flutter-framework-code.

I think if you want to use them, you'd have to register a listener an them and in the listener you'd do something like

```dart
void handleNewNetry(String entry) {
  setState(() {
    _entries.add(entry);
  });
}
```

But then you'll also have to figure out when to remove the listener (i.e. close the StreamSubscription).

If I had to design this API, I would probably make `find` return a `ValueNotifier` that notifies you whenever a new entry has been added to its `value`. You'd still have to do the listener thing, though.

Reply:
Yea, that makes sense.  Shutting down the stream or ValueNotifier when it is interrupted seems like it would make the example much more complicated.  Maybe I'll keep it like this since it's already complicated enough.